### ISSUE TITLE
Statistics for Evaluating Machine Learning Models

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#457]
Statistics for Evaluating Machine Learning Models<br>
<br>
https://ift.tt/2yWGWgI<br>
<br>
Jason Brownlee<br>
<br>
<p>Tom Mitchell&rsquo;s classic 1997 book &ldquo;<em>Machine Learning</em>&rdquo; provides a chapter dedicated to statistical methods for evaluating machine learning models.</p>
<p>Statistics provides an important set of tools used at each step of a machine learning project. A practitioner cannot effectively evaluate the skill of a machine learning model without using statistical methods. Unfortunately, statistics is an area that is foreign to most developers and computer science graduates. This makes the chapter in Mitchell&rsquo;s seminal machine learning text an important, if not required, reading by practitioners.</p>
<p>In this post, you will discover statistical methods recommended by Mitchel to evaluate and compare machine learning models.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>How to calculate confidence intervals for classification accuracy or error.</li>
<li>The statistical foundation for calculating confidence intervals.</li>
<li>How to generally compare the performance of machine learning algorithms.</li>
</ul>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/07/Statistics-for-Evaluating-Machine-Learning-Models.jpg" alt="Statistics for Evaluating Machine Learning Models" width="640" class="wp-image-5688" height="320">
<p class="wp-caption-text">Statistics for Evaluating Machine Learning Models<br>
Photo by <a href="https://www.flickr.com/photos/equinoxefr/1815074519/">Pierre (Rennes)</a>, some rights reserved.</p>
</div>
<h2>Machine Learning Book</h2>
<p><a href="http://www.cs.cmu.edu/~tom/">Tom Mitchell</a> wrote what might be the classic textbook on applied machine learning, titled &ldquo;<a href="https://amzn.to/2Jja7vF">Machine Learning</a>&rdquo; and released in 1997.</p>
<a href="http://www.amazon.com/dp/0071154671?tag=inspiredalgor-20"><img alt="Amazon Image" width="333" src="https://images-na.ssl-images-amazon.com/images/I/312t3q1DwtL.jpg" height="500"></a>
<p>In the book he dedicated an entire chapter to the statistical methods required to evaluate machine learning models and algorithms. Specifically, Chapter 5 titled &ldquo;<em>Evaluating Hypotheses</em>&ldquo;.</p>
<blockquote>
<p>Chapter 5 presents basic concepts from statistics and estimation theory, focusing on evaluating the accuracy of hypotheses using limited samples of data. This includes the calculation of confidence intervals for estimating hypothesis accuracy and methods for comparing the accuracy of learning methods.</p>
</blockquote>
<p>&mdash; Page 16, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<p>In this post, we are going to take a closer look at this chapter and review the statistical methods recommended by Mitchell at that time. The fundamentals of statistical methods do not change and may be just as useful and relevant now, 20 years since the publication of the book.</p>
<p>It is important to note that when Mitchell refers to hypotheses, he is referring to learned models, the results of running a learning algorithm on a dataset. Evaluating and comparing hypotheses means comparing learned models, which is different from evaluating and comparing machine learning algorithms, which could be trained on different samples from the same problem or different problems.</p>
<p>Chapter 5 on evaluating hypotheses is divided into 7 parts; they are as follows:</p>
<ul>
<li>5.1. Motivation</li>
<li>5.2. Estimating Hypothesis Accuracy</li>
<li>5.3. Basics of Sampling Theory</li>
<li>5.4. A General Approach for Deriving Confidence Intervals</li>
<li>5.5. Difference in Error of Two Hypotheses</li>
<li>5.6. Comparing Learning Algorithms</li>
<li>5.7. Summary and Further Reading</li>
</ul>
<p>We will spend time looking at each section and summarize the statistical methods and recommendations made.</p>
<center>
<h3>Need help with Statistics for Machine Learning?</h3>
<p>Take my free 7-day email crash course now (with sample code).</p>
<p>Click to sign-up and also get a free PDF Ebook version of the course.</p>
<p><a href="https://machinelearningmastery.lpages.co/leadbox/142f75173f72a2%3A164f8be4f346dc/5750943224168448/" target="_blank">Download Your FREE Mini-Course</a></p>
</center>
<h2>Motivation</h2>
<p>The chapter starts off by stating the importance of evaluating hypotheses in machine learning.</p>
<blockquote>
<p>Empirically evaluating the accuracy of hypotheses is fundamental to machine learning.</p>
</blockquote>
<p>&mdash; Page 128, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<p>The chapter is motivated by three questions; they are:</p>
<ul>
<li>Given the observed accuracy of a hypothesis over a limited sample of data, how well does this estimate its accuracy over additional examples?</li>
<li>Given that one hypothesis outperforms another over some sample of data, how probable is it that this hypothesis is more accurate in general?</li>
<li>When data is limited, what is the best way to use this data to both learn a hypothesis and estimate its accuracy?</li>
</ul>
<p>All three questions are closely related.</p>
<p>The first question raises concerns over errors in the estimate of model skill and motivates the need for confidence intervals.</p>
<p>The second question raises concerns about making decisions based on model skill on small samples and motivates statistical hypothesis testing.</p>
<p>And finally, the third question considers the economical use of small samples and motivates resampling methods such as k-fold cross-validation.</p>
<blockquote>
<p>This chapter discusses methods for evaluating learned hypotheses, methods for comparing the accuracy of two hypotheses, and methods for comparing the accuracy of two learning algorithms when only limited data is available.</p>
</blockquote>
<p>&mdash; Page 129, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<p>The motivation closes with a reminder of the difficulty of estimating the skill of a hypothesis. Specifically the introduction of bias and variance in the estimate of the skill of a model:</p>
<ul>
<li>
<strong>Bias in the estimate</strong>. When the model is evaluated on training data, overcome by evaluating the model on a hold out test set.</li>
<li>
<strong>Variance in the estimate</strong>. When the model is evaluated on an independent test set, overcome by using larger tests sets.</li>
</ul>
<h2>Estimating Hypothesis Accuracy</h2>
<p>The skill or prediction error of a model must be estimated, and as an estimate, it will contain error.</p>
<p>This is made clear by distinguishing between the true error of a model and the estimated or sample error.</p>
<blockquote>
<p>One is the error rate of the hypothesis over the sample of data that is available. The other is the error rate of the hypothesis over the entire unknown distribution D of examples.</p>
</blockquote>
<p>&mdash; Page 130, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<ul>
<li>
<strong>Sample Error</strong>. Estimate of true error calculated on a data sample.</li>
<li>
<strong>True Error</strong>. Probability that a model will misclassify a randomly selected example from the domain.</li>
</ul>
<p>We want to know the true error, but we must work with the estimate, approximated from a data sample. This raises the question of how good is a given estimate of error?</p>
<p>One approach is to calculate a confidence interval around the sample error that is large enough to cover the true error with a very high likelihood, such as 95%.</p>
<p>Assuming that the error measure is a discrete proportion, such as classification error, the calculation of the confidence interval is calculated as follows:</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/04/Calculation-of-Confidence-Interval-for-Classification-Error.png" alt="Calculation of Confidence Interval for Classification Error" width="908" class="wp-image-5684" height="164">
<p class="wp-caption-text">Calculation of Confidence Interval for Classification Error<br>
Taken from Machine Learning, 1997.</p>
</div>
<p>Where <em>error_s</em> is the sample error, <em>n</em> is the total number of instances in the test set used to calculate the sample error and 1.96 is the critical value from the Gaussian distribution for a likelihood of 95%.</p>
<h2>Basics of Sampling Theory</h2>
<p>The equation to calculate the confidence interval from the previous section makes many assumptions.</p>
<p>This section works through these assumptions in order to provide a basis of understanding for the calculation.</p>
<blockquote>
<p>This section introduces basic notions from statistics and sampling theory, including probability distributions, expected value, variance, Binomial and Normal distributions, and two-sided and one-sided intervals.</p>
</blockquote>
<p>&mdash; Page 132, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<p>Usefully, a table is provided that summarizes the key concepts of this section, provided below.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/04/Summary-of-Key-Statistical-Concepts-for-Calculating-a-Confidence-Interval-1024x760.png" alt="Summary of Key Statistical Concepts for Calculating a Confidence Interval" width="1024" class="wp-image-5685" height="760">
<p class="wp-caption-text">Summary of Key Statistical Concepts for Calculating a Confidence Interval<br>
Taken from Machine Learning, 1997.</p>
</div>
<p>This section provides an important statistical foundation for the confidence interval and is recommended reading.</p>
<p>To save repeating all of this theory, the crux of the section is as follows:</p>
<ul>
<li>Proportional values like classification accuracy and classification error fit a Binomial distribution.</li>
<li>The Binomial distribution characterizes the probability of a binary event, such as a coin flip or a correct/incorrect classification prediction.</li>
<li>The mean is the expected value in the distribution, the variance is the average distance that samples have from the mean, and the standard deviation is the variance normalized by the size of the data sample.</li>
<li>Ideally, we seek an unbiased estimate of our desired parameter that has the smallest variance.</li>
<li>Confidence intervals provide a way to quantify the uncertainty in a population parameter, such as a mean.</li>
<li>The Binomial distribution can be approximated with the simpler Gaussian distribution for large sample sizes, e.g. 30 or more observations.</li>
<li>The interval can be centered on the mean called two-sided, but can also be one-sided, such as a radius to the left or right of the mean.</li>
</ul>
<h2>A General Approach for Deriving Confidence Intervals</h2>
<p>Given the equation to calculate the confidence intervals for proportional values and the statistical reasoning behind the equation, a general procedure is presented to calculate confidence intervals.</p>
<p>The procedure is summarized below.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/04/General-Procedure-for-Calculating-Confidence-Intervals-1024x322.png" alt="General Procedure for Calculating Confidence Intervals" width="1024" class="wp-image-5686" height="322">
<p class="wp-caption-text">General Procedure for Calculating Confidence Intervals<br>
Taken from Machine Learning, 1997.</p>
</div>
<p>The central limit theorem is also presented. It can be summarized by the finding that the sum (or normalized sum in the case of the mean) of independent observations will represent a sample from a Gaussian distribution. For example, the mean skill of a model on different independent data samples will be Gaussian.</p>
<p>This is an invaluable finding as we know so much about the Gaussian and can comment on the likelihood of two samples (mean values) belonging to the same or different Gaussian distributions, such as in the case of the skill of machine learning algorithms.</p>
<blockquote>
<p>The Central Limit Theorem is a very useful fact because it implies that whenever we define an estimator that is the mean of some sample (e.g., errors(h) is the mean error), the distribution governing this estimator can be approximated by a Normal distribution for sufficiently large n.</p>
</blockquote>
<p>&mdash; Page 143, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<h2>Difference in Error of Two Hypotheses</h2>
<p>This section looks at applying the general procedure for calculating confidence intervals to the estimated difference in classification error between two models.</p>
<p>The approach assumes that each model was trained on a different independent sample of the data. Therefore, the calculation of the confidence interval in the difference in error between the two models adds the variance from each model.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/04/Confidence-Interval-for-the-Difference-in-Error-Between-Two-Models-1024x130.png" alt="Confidence Interval for the Difference in Error Between Two Models" width="1024" class="wp-image-5687" height="130">
<p class="wp-caption-text">Confidence Interval for the Difference in Error Between Two Models<br>
Taken from Machine Learning, 1997.</p>
</div>
<p>This section also introduces the idea of statistical hypothesis testing as an alternative to calculating confidence intervals.</p>
<blockquote>
<p>In some cases we are interested in the probability that some specific conjecture is true, rather than in confidence intervals for some parameter.</p>
</blockquote>
<p>&mdash; Page 145, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<p>Interestingly, the topic is motivated by the likelihood of a one-sided confidence interval containing the true estimate of error for a model in order to determine the probability of one model being better than another.</p>
<p>I found this explanation less clear than it should have been.</p>
<h2>Comparing Learning Algorithms</h2>
<p>This final content section of the chapter focuses on the comparison of machine learning algorithms.</p>
<p>This is different from comparing models (hypotheses) as comparing algorithms involves training them and evaluating them potentially on multiple different samples of data from the domain.</p>
<p>The comparison of two algorithms is motivated by estimating the expected or mean difference between the two methods. A procedure is presented that uses k-fold cross-validation where each algorithm is trained and evaluated on the same splits of the data. A final mean difference in error is calculated, from which a confidence interval can be estimated.</p>
<p>The calculation of the confidence interval is updated to account for the reduced number of degrees of freedom as each algorithm is evaluated on the same test set.</p>
<p>The paired Student&rsquo;s t-test is introduced as a statistical hypothesis test for quantifying the likelihood that two means belong to the same (or different) distributions. This test can be used with the outlined procedure, but only if each train and test set contain independent samples, a fact that is not the case with default k-fold cross-validation.</p>
<blockquote>
<p>In particular, in this idealized method we modify the procedure of Table 5.5 so that on each iteration through the loop it generates a new random training set Si and new random test set Ti by drawing from this underlying instance distribution instead of drawing from the fixed sample Do</p>
</blockquote>
<p>&mdash; Page 148, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<p>The section ends by outlining practical considerations when comparing machine learning algorithms.</p>
<p>Mitchell reminds us that the Student&rsquo;s t-test does not technically apply in the case where we use resampling methods. Nevertheless, he recommends using k-fold cross-validation or random sampling in order to estimate the variance in the estimate of model error, as they are the only methods available.</p>
<p>This is less than ideal as the expectations of the statistical test will be violated, increasing type I errors.</p>
<blockquote>
<p>It is wise to keep in mind that statistical models rarely fit perfectly the practical constraints in testing learning algorithms when available data is limited. Nevertheless, they do provide approximate confidence intervals that can be of great help in interpreting experimental comparisons of learning methods.</p>
</blockquote>
<p>&mdash; Page 150, <a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</p>
<h2>Summary and Further Reading</h2>
<p>The chapter ends with a summary of the main points hit, exercises that can be completed to confirm an understanding of the equations, and a great list of references and further reading.</p>
<p>Briefly, the main points are:</p>
<ul>
<li>Confidence intervals can be used to quantify the uncertainty in model error.</li>
<li>The calculation of the confidence intervals importantly depends on the distribution of the quantity being estimated.</li>
<li>A cause of errors in the evaluation of model skill is estimation bias, evaluating a model on the training dataset will introduce an optimistic bias.</li>
<li>Another cause of errors in the evaluation of model skill is estimation variance, which can be decreased by increasing the hold out test size.</li>
<li>Resampling with the k-fold cross-validation procedure provides an effective way to compare machine learning algorithms.</li>
<li>When using statistical methods, we must make assumptions, such as assumptions around the distribution of parameters. Even with confidence intervals, we don&rsquo;t know what is true, just what is likely.</li>
</ul>
<p>Interestingly, two technical reports by <a href="https://en.wikipedia.org/wiki/Thomas_G._Dietterich">Thomas Dietterich</a> are referenced. Dietterich went on to publish the important 1998 paper titled &ldquo;<a href="http://sci2s.ugr.es/keel/pdf/algorithm/articulo/dietterich1998.pdf">Approximate Statistical Tests for Comparing Supervised Classification Learning Algorithms</a>&rdquo; that describes the unreliability of the paired Student&rsquo;s t-test when comparing machine learning algorithms with random resampling and k-fold cross-validation.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<ul>
<li>
<a href="https://amzn.to/2Jja7vF">Machine Learning</a>, 1997.</li>
<li><a href="http://www.cs.cmu.edu/~tom/">Tom Mitchell Homepage</a></li>
<li><a href="https://en.wikipedia.org/wiki/Tom_M._Mitchell">Tom M. Mitchell on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Thomas_G._Dietterich">Thomas Dietterich on Wikipedia</a></li>
<li>
<a href="http://web.cs.iastate.edu/~honavar/dietterich98approximate.pdf">Approximate Statistical Tests for Comparing Supervised Classification Learning Algorithms</a>, 1998.</li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered statistical methods to evaluate and compare machine learning models.</p>
<p>Specifically, you learned:</p>
<ul>
<li>How to calculate confidence intervals for classification accuracy or error.</li>
<li>The statistical foundation for calculating confidence intervals</li>
<li>How to generally compare the performance of machine learning algorithms.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/statistics-for-evaluating-machine-learning-models/">Statistics for Evaluating Machine Learning Models</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
July 2, 2018 at 03:08AM
